### PR TITLE
SqlClient improve managed memory usage

### PR DESCRIPTION
**Pull request by Wraith2**
The managed Tds implementation is only used on Unix where the native implementation cannot be used, this is for performance reasons. I've investigated and made some changes which improve the performance of the managed mode. There is a still a lot that can be done but I think they'll need to build on this start.

The release build now works so I can get inproc BDN numbers:

|      Method |     Mean |    Error |   StdDev |   Median | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------ |---------:|---------:|---------:|---------:|------------:|------------:|------------:|--------------------:|
| SyncFloat32 | 235.2 ms | 4.871 ms | 12.31 ms | 228.8 ms |   3000.0000 |           - |           - |           593.75 KB |

|      Method |     Mean |    Error |   StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------------ |---------:|---------:|---------:|------------:|------------:|------------:|--------------------:|
| SyncFloat32 | 229.9 ms | 3.993 ms | 3.540 ms |           - |           - |           - |           281.25 KB |

Which shows that there is a large improvement in memory usage. The numbers are a bit sterile though. Compare profiles from before where there is a gen0 roughly every 100ms.

before

to after:

after

The changes are straightforward in terms of code changes even though it was tricky working out the right places to make them. Packets were being dropped without disposing them causing their rented arrays to be lost, this meant a lot of churn in array allocations. The mars mux header writing code was allocating a lot of temporary arrays so I replaced that code with a leaner version and stopped re-allocation of the header itself where it could be re-used.

The changes pass standard and manual sql tests.

/cc all the usual people, @saurabh500 @AfsanehR @keeratsingh

### CODE REVIEW

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs [hunk 1]
```diff
@@ -44,6 +44,46 @@ internal class SNISMUXHeader
         public uint length;
         public uint sequenceNumber;
         public uint highwater;
+
+        public void Read(ReadOnlySpan<byte> bytes)
+        {
+            SMID = bytes[0];
+            flags = bytes[1];
+            sessionId = BitConverter.ToUInt16(bytes.Slice(2,2));
```
Comment (Tornhoof) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs]: May I suggest slicing into a temp variable and use that one? Then you don't need to manually track the start offset, makes it easier to read.

Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs]: The only way I can think to do that seems less clear to me, can you give an example?

**Comment by saurabh500** [comment]
How did you test this?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs (hunk 1 above)
Comment (Tornhoof) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs]: Sorry. My bad, I've read your code wrong, my suggestion doesn't work.

**Comment by Wraith2** [comment]
I identified all the tests that break with the assertion covered in corefx/33930 and marked them with activeissue since there are only 12 of them. I could also have run the manual tests in release mode but that way I wouldn't have seen any other assertions and I think the diagnostics of a failure would have been helpful.

**Comment by saurabh500** [comment]
All on Windows with switch flipped? Can you list the tests that are disabled? 

What does you benchmark test?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs (hunk 1 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs]: Thanks for looking it over 😁

**Comment by Wraith2** [comment]
Yes, all windows with the switch flipped, My local version lets me set it in both release and managed mode but otherwise it's identical. 

The benchmark is a single connection, single command, sproc insert of a float into a table. [code is here](https://github.com/Wraith2/SqlBench/blob/1e26aee93bc81c008bc51581537028b492b39ad7/Program.cs#L196). The intention is just to test simple small packets passing back and forth.

I've added the list of disabled tests to https://github.com/dotnet/corefx/issues/33930 . The Tvp test is also not unix specific and this is covered in https://github.com/dotnet/corefx/issues/27858

**Comment by saurabh500** [comment]
The code changes are in MARS data exchange. Have you been enabling Mars in your connection string for benchmark?

**Comment by Wraith2** [comment]
Yes. And there are only two mars tests disabled, all the others pass. The hard work of muxing is all untouched I've only changed how the bytes are read/written to the buffer slightly and moved that logic into the smux header class.

**Comment by Wraith2** [comment]
@dotnet-bot test Linux arm64 Release Build
@dotnet-bot test Packaging All Configurations x64 Debug Build

**Comment by saurabh500** [comment]
This is not a transient failure in CI. There are compilation errors.

**Comment by Wraith2** [comment]
Ah, I see them now in the packaging version. Just BitConverter overload stuff which is easily fixed. Thanks for pointing it out.

**Comment by Wraith2** [comment]
Using this basis I have a branch which makes further improvements.
new

**Comment by saurabh500** [comment]
Sounds good @Wraith2 
The team is dealing with some post-holiday catch up internally. These PRs haven't been forgotten. 
Will get to these after taking care of some of the backlog. I am hoping to get the review cadence going, starting tomorrow.

**Comment by Wraith2** [comment]
I'm not back at work yet, still in holiday mode. I've only just reached the point where my new changes are stable so I thought I should get some numbers.

**Comment by Wraith2** [comment]
@dotnet-bot test Linux x64 Release Build

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsConnection.cs [hunk 2]
```diff
@@ -21,7 +21,7 @@ internal class SNIMarsConnection
         private ushort _nextSessionId = 0;
         private int _currentHeaderByteCount = 0;
         private int _dataBytesLeft = 0;
-        private SNISMUXHeader _currentHeader;
+        private SNISMUXHeader _currentHeader = new SNISMUXHeader();
```
Comment (corivera) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsConnection.cs]: readonly?

Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsConnection.cs]: Sure, done.

**Comment by AfsanehR** [comment]
@dotnet-bot Windows x64 Debug Build
@dotnet-bot Windows x86 Release Build

**Comment by AfsanehR** [comment]
@dotnet-bot test Windows x64 Debug Build
@dotnet-bot test Windows x86 Release Build

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs [hunk 3]
```diff
@@ -44,6 +44,46 @@ internal class SNISMUXHeader
         public uint length;
         public uint sequenceNumber;
         public uint highwater;
+
+        public void Read(byte[] bytes)
+        {
+            SMID = bytes[0];
+            flags = bytes[1];
```
Comment (Poltuu) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs]: would inverting line 51 and 50 here allow for the same benefit discussed line 61 ?
> access the highest element first to cause the largest range check in the jit

Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNICommon.cs]: Yes, I decided it was unlikely to make a significant difference by the time i'd got to this one though. If later benchmarking shows this as a hot path i'll revisit and do the same, in the meantime the JIT constantly improves and may no longer need the reverse order layout to produce optimal code. We'll see.

**Pull request merged**